### ISSUE TITLE
Parameter missing from dataclass __init__ if defined inside a conditional context

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#3422]
Consider the following example:

```python
import dataclasses
from typing import TYPE_CHECKING


@dataclasses.dataclass
class Foo:
    if TYPE_CHECKING:
        a: str | int
    else:
        a: str

    b: str


f = Foo(a="xxx", b="yyy")


def get_foo() -> Foo:
    ...


foo = get_foo()
reveal_type(foo.b)
reveal_type(foo.a)
```

Running pyright against this gives:
```
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/xxx.py
  /tmp/xxx.py:15:9 - error: No parameter named "a" (reportGeneralTypeIssues)
  /tmp/xxx.py:23:13 - information: Type of "foo.b" is "str"
  /tmp/xxx.py:24:13 - information: Type of "foo.a" is "str | int"
1 error, 0 warnings, 2 informations 
Completed in 0.612sec
```

Pyright is correctly using the definition inside the `if TYPE_CHECKING` on the reveal_type, but it is is not considering it for the dataclass `__init__` param.

The same happens for any conditional. For example, changing `if TYPE_CHECKING` to `if True` will make pyright consider a to be `str | int` and `if False` to consider it to be `str`, but in both cases the parameter will also be missing from `__init__`.

**Comment by erictraut** [comment]
Hmm, I don't know if it's worth supporting this edge case. What you're doing here is pretty suspect.

I'll investigate it further to see whether it's relatively easy to accommodate this case.

**Comment by bellini666** [comment]
> What you're doing here is pretty suspect.

This is just a minimal example for reproduction. But for contextualization, the use case I had is from a lib that needs the type to be typed as a specific type at runtime (because it parses and uses the type for serialization/deserialization), but it is actually fine for me to pass other types there because I handle normalize it in `__post_init__`.

**Comment by hmc-cs-mdrissi** [comment]
Hmm, I've had a similar use case for serialization/deserialization. I use something like Annotated[type1, type2] where type2 is runtime type for situations where serialization type allowed differs from type checking type allowed.

A second alternative that works for situations like this,

```python
if TYPE_CHECKING:
  alias: TypeAlias = "str | int"
else:
  alias: TypeAlias = str

@dataclasses.dataclass
class Foo:
  a:  alias
  b: str
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.244, which I just published. It will also be included in the next release of pylance.
